File: ios/Synapse/Core/Library/ResourceFileStore.swift
import Foundation
import Observation

/// Downloading and keeping the source documents.
///
/// These are textbooks — tens of megabytes each — so they are streamed to disk
/// rather than held in memory, and kept once fetched. A student who downloads
/// an anatomy PDF on wifi should still have it on the ward, which is the whole
/// reason the app exists rather than the website.
@MainActor
@Observable
final class ResourceFileStore {

    enum State: Equatable {
        case notDownloaded
        case downloading(fraction: Double)
        case ready(URL)
        case failed(String)
    }

    private(set) var states: [String: State] = [:]

    private let api: SynapseAPI
    private var tasks: [String: Task<Void, Never>] = [:]

    init(api: SynapseAPI) {
        self.api = api
    }

    func state(for id: String) -> State {
        if let known = states[id] { return known }
        // A file already on disk from a previous session.
        if let url = Self.existingFile(id) {
            states[id] = .ready(url)
            return .ready(url)
        }
        return .notDownloaded
    }

    /// Where a resource lives once downloaded.
    ///
    /// Application Support rather than Caches: the system may evict Caches
    /// under disk pressure, and a textbook silently disappearing the night
    /// before an exam is exactly the failure this app is meant to prevent.
    static func directory() throws -> URL {
        let base = try FileManager.default.url(
            for: .applicationSupportDirectory, in: .userDomainMask,
            appropriateFor: nil, create: true
        )
        let folder = base.appendingPathComponent("Synapse/Resources", isDirectory: true)
        try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        return folder
    }

    static func fileURL(_ id: String) -> URL? {
        // The ID comes from authored content, so it is not trusted as a path
        // component: anything but the expected shape is refused rather than
        // allowed to escape the directory.
        let safe = id.filter { $0.isLetter || $0.isNumber || $0 == "_" || $0 == "-" }
        guard safe == id, !safe.isEmpty else { return nil }
        return try? directory().appendingPathComponent("\(safe).pdf")
    }

    static func existingFile(_ id: String) -> URL? {
        guard let url = fileURL(id), FileManager.default.fileExists(atPath: url.path) else { return nil }
        return url
    }

    func download(_ id: String) {
        guard tasks[id] == nil else { return }
        if case .ready = state(for: id) { return }

        states[id] = .downloading(fraction: 0)
        tasks[id] = Task { [weak self] in
            guard let self else { return }
            defer { Task { @MainActor in self.tasks[id] = nil } }

            do {
                guard let destination = Self.fileURL(id) else {
                    await MainActor.run { self.states[id] = .failed("That resource has an unusable identifier.") }
                    return
                }
                let temporary = try await api.downloadResource(id: id) { fraction in
                    Task { @MainActor in self.states[id] = .downloading(fraction: fraction) }
                }
                // Move rather than copy, and replace any half-file from a
                // previous failed attempt.
                try? FileManager.default.removeItem(at: destination)
                try FileManager.default.moveItem(at: temporary, to: destination)

                var resourceValues = URLResourceValues()
                resourceValues.isExcludedFromBackup = true
                var mutable = destination
                try? mutable.setResourceValues(resourceValues)

                await MainActor.run { self.states[id] = .ready(destination) }
            } catch APIError.notFound {
                await MainActor.run {
                    self.states[id] = .failed("This resource has not been uploaded yet.")
                }
            } catch {
                await MainActor.run {
                    self.states[id] = .failed("Could not download it. Check your connection and try again.")
                }
            }
        }
    }

    func cancel(_ id: String) {
        tasks[id]?.cancel()
        tasks[id] = nil
        states[id] = .notDownloaded
    }

    func delete(_ id: String) {
        guard let url = Self.fileURL(id) else { return }
        try? FileManager.default.removeItem(at: url)
        states[id] = .notDownloaded
    }

    /// What the downloaded documents are costing in storage.
    static func bytesOnDisk() -> Int64 {
        guard let folder = try? directory(),
              let entries = try? FileManager.default.contentsOfDirectory(
                at: folder, includingPropertiesForKeys: [.fileSizeKey]
              )
        else { return 0 }

        return entries.reduce(0) { total, url in
            let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
            return total + Int64(size)
        }
    }
}

File: android/app/src/main/java/com/nishany/android/core/library/ResourceFileStore.kt
package com.nishany.android.core.library

import android.content.Context
import com.nishany.android.core.api.ApiError
import com.nishany.android.core.api.NishanyApi
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Downloading and keeping the source PDFs behind library resources.
 *
 * A port of iOS `Core/Library/ResourceFileStore.swift`. These are textbooks
 * -- tens of megabytes each -- so they are streamed to disk via
 * [NishanyApi.downloadResourceFile] rather than held in memory, and kept once
 * fetched: a student who downloads an anatomy PDF on wifi should still have
 * it on the ward.
 *
 * Held once in [com.nishany.android.AppGraph], not per-screen: an in-flight
 * download must survive the student navigating away from the reader and back
 * (unlike a ViewModel, which the nav graph tears down on that same trip).
 */
class ResourceFileStore(
    private val api: NishanyApi,
    private val filesDir: File,
) {
    sealed interface State {
        data object NotDownloaded : State
        data class Downloading(val fraction: Float?) : State
        data class Ready(val file: File) : State
        data class Failed(val message: String) : State
    }

    // ponytail: SupervisorJob rather than per-download cancellation
    // plumbing through a UI layer -- [cancel] below tears down and restarts
    // this scope's children directly, which is all a "cancel this download"
    // button needs.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val jobs = mutableMapOf<String, Job>()
    private val _states = MutableStateFlow<Map<String, State>>(emptyMap())
    val states: StateFlow<Map<String, State>> = _states.asStateFlow()

    /** Where a resource lives once downloaded. Not `context.cacheDir`: the system may evict that under disk pressure, and a textbook disappearing the night before an exam is exactly what this store exists to prevent. */
    private fun directory(): File = File(filesDir, "resources").apply { mkdirs() }

    /** The id comes from authored content, so it is not trusted as a path component -- anything but the expected shape is refused rather than allowed to escape the directory. */
    private fun fileFor(id: String): File? {
        val safe = id.filter { it.isLetterOrDigit() || it == '_' || it == '-' }
        if (safe != id || safe.isEmpty()) return null
        return File(directory(), "$safe.pdf")
    }

    fun state(id: String): State {
        _states.value[id]?.let { return it }
        val existing = fileFor(id)?.takeIf { it.exists() }
        if (existing != null) {
            _states.update { it + (id to State.Ready(existing)) }
            return State.Ready(existing)
        }
        return State.NotDownloaded
    }

    fun download(id: String) {
        if (jobs[id] != null) return
        if (state(id) is State.Ready) return

        val destination = fileFor(id)
        if (destination == null) {
            _states.update { it + (id to State.Failed("That resource has an unusable identifier.")) }
            return
        }

        _states.update { it + (id to State.Downloading(fraction = null)) }
        jobs[id] = scope.launch {
            try {
                api.downloadResourceFile(id, destination) { fraction ->
                    _states.update { it + (id to State.Downloading(fraction)) }
                }
                _states.update { it + (id to State.Ready(destination)) }
            } catch (e: ApiError.NotFound) {
                _states.update { it + (id to State.Failed("This resource has not been uploaded yet.")) }
            } catch (e: Exception) {
                _states.update { it + (id to State.Failed("Could not download it. Check your connection and try again.")) }
            } finally {
                jobs.remove(id)
            }
        }
    }

    fun cancel(id: String) {
        jobs.remove(id)?.cancel()
        _states.update { it + (id to State.NotDownloaded) }
    }
}
